site/content/posts/2023-08-09-v0.25.0-impersonation-proxy-with-external-certs.md

```diff
@@ -0,0 +1,337 @@
+$ kubectl get credentialissuers pinniped-concierge-config \
+    --output yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+...
+  spec:
+    impersonationProxy:
+      externalEndpoint: impersonation-proxy-mtls.local
+      mode: enabled
+      service:
+        annotations:
+          # Ignore any annotations
+        type: ClusterIP
+      tls:
+        certificateAuthorityData: LS0tLUJFR0l..........
+        secretName: impersonation-proxy-tls-serving-cert
+...
+
+# Confirm that the ClusterIP service for the impersonation proxy was automatically created (may take a minute)
+$ kubectl get service pinniped-concierge-impersonation-proxy-cluster-ip \
+    --namespace pinniped-concierge \
+    --output yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+
+# Configure a webhook authenticator to tell Concierge to validate static tokens using the installed local-user-authenticator
+$ cat << EOF > concierge.webhookauthenticator.yaml
+apiVersion: authentication.concierge.pinniped.dev/v1alpha1
+kind: WebhookAuthenticator
+metadata:
+  name: local-user-authenticator
+spec:
+  endpoint: https://local-user-authenticator.local-user-authenticator.svc/authenticate
+  tls:
+    certificateAuthorityData: $(cat local-user-authenticator-ca.pem.b64)
+EOF
+
+# Create the webhook authenticator
+$ kubectl apply \
+    --filename concierge.webhookauthenticator.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+```
+
+Now deploy a Contour `HTTPProxy` ingress that fronts the `ClusterIP` service for the impersonation proxy.
+
+We need to use TLS passthrough in this case, so that the client (kubectl and the pinniped CLI) can establish TLS directly
+with the impersonation proxy, and so that client certs used for mTLS will be sent to the impersonation proxy.
+
+Note in particular the `spec.tcpproxy` block, which is different than the typical `spec.rules` block. 
+`spec.tcpproxy` is required when using `spec.virtualhost.tls.passthrough: true`. 
+
+See https://projectcontour.io/docs/1.25/config/tls-termination/#tls-session-passthrough for more details.
+
+```shell
+$ cat << EOF > contour-ingress-impersonation-proxy.yaml
+---
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: impersonation-proxy
+  namespace: pinniped-concierge
+spec:
+  virtualhost:
+    fqdn: impersonation-proxy-mtls.local
+    tls:
+      passthrough: true
+  tcpproxy:
+    services:
+    - name: pinniped-concierge-impersonation-proxy-cluster-ip
+      port: 443
+EOF
+
+$ kubectl apply \
+    --filename contour-ingress-impersonation-proxy.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+```
+
+Now generate the Pinniped kubeconfig so that you can perform mTLS with the impersonation proxy.
+
+Note that using a static-token does embed those credentials into your kubeconfig.
+Never use `local-user-authenticator` in production.
+
+```shell
+# add 127.0.0.1 impersonation-proxy-mtls.local to your /etc/hosts!
```

Comment:
I think 

```shell
# comments like this are nice to know
```
We should perhaps bump this up to a "you must do this in order for it to work" kind of entry. 🤔

Reply:
Since we are accessing a `kind` cluster running on our local machine, we will need some way to ensure HTTP requests issued loop back to localhost.  An easy way to do this is via the `/etc/hosts` file.

Reply:
resolving locally